pyscripts/build_docker_image.py

```diff
@@ -10,193 +10,237 @@
-    "password": os.environ["IOTHUB_E2E_REPO_PASSWORD"],
-}
-
-
-def get_dockerfile_directory(tags):
-    script_dir = os.path.dirname(os.path.realpath(__file__))
-    return os.path.normpath(os.path.join(script_dir, "../ci-wrappers/" + tags.language))
-
-
-def print_filtered_docker_line(line):
-    try:
-        obj = json.loads(line.decode("utf-8"))
-    except:
-        print(line)
-    else:
-        if "status" in obj:
-            if "id" in obj:
-                if obj["status"] not in [
-                    "Waiting",
-                    "Downloading",
-                    "Verifying Checksum",
-                    "Extracting",
-                    "Preparing",
-                    "Pushing",
-                ]:
-                    print("{}: {}".format(obj["status"], obj["id"]))
+class BuildDockerImage:
+
+    def __init__(self, args):
+
+        colorama.init(autoreset=True)
+
+        default_repo = "(Azure/azure-iot-sdk-BLAH)"
+        all_languages = ["c", "csharp", "python", "pythonpreview", "node", "java"]
+
+        parser = argparse.ArgumentParser(description="build docker image for testing")
+        parser.add_argument(
+            "--language",
+            help="language to build",
+            type=str,
+            required=True,
+            choices=all_languages,
+        )
+        parser.add_argument("--repo", help="repo with source", type=str, default=default_repo)
+        parser.add_argument(
+            "--commit", help="commit to apply (ref or branch)", type=str, default="master"
+        )
+        parser.add_argument(
+            "--variant", help="Docker image variant (blank for default)", type=str, nargs="?", const=""
+        )
+        args = parser.parse_args()
+
+        if args.repo == default_repo:
+            if args.language == "pythonpreview":
+                args.repo = "Azure/azure-iot-sdk-python-preview"
+            else:
+                args.repo = "Azure/azure-iot-sdk-" + args.language
+            print(Fore.YELLOW + "Repo not specified.  Defaulting to " + args.repo)
+
+        print_separator = "".join("/\\" for _ in range(80))
+
+        auth_config = {
+            "username": os.environ["IOTHUB_E2E_REPO_USER"],
+            "password": os.environ["IOTHUB_E2E_REPO_PASSWORD"],
+        }
+
+
+        def get_dockerfile_directory(tags):
+            script_dir = os.path.dirname(os.path.realpath(__file__))
+            return os.path.normpath(os.path.join(script_dir, "../ci-wrappers/" + tags.language))
+
+
+        def print_filtered_docker_line(lineobj):
+            lineobj = lineobj.splitlines()
+            for line in lineobj:
+                try:
+                    linecode = line.strip().decode("utf-8")
+                    linecode = linecode[:linecode.rfind("\r\n")] + '}'
+                    jline = json.loads(linecode)
+                    obj = jline["stream"]
+                except:
+                    print(''.join([chr(i) if i < 128 else '\x23' for i in line]))
```

Comment:
>print(''.join([chr(i) if i < 128 else '\x23' for i in line])) [](start = 20, length = 61)

reading this, I'm still not sure if line is a string or a buffer.  It looks like it's a buffer here, and it looks like you're converting to 7-bit ascii and replacing everything that isn't 7-bit ascii with a '#'.  Huh?  Why?

Reply:
Bert - I can totally see the errors and confusion here.  This as an attempt to fix the invalid json format returned by docker.APIClient in some cases - the json that was returned by docker.APIClient was not valid json and passing it to json.loads() threw an exception. The json was missing the closing curley.

Yes, I got a little overzealous with the cut and paste from the stackoverflow "solution", and yes it is double decoded into Ascii7.
I will revisit and fix this code. I think the best solution here is to just catch the exception from json.loads() and print the Ascii7 rather than trying to fix it.

Reply:
The best solution here may be to catch the exception from json.loads() and print the invalid json as Ascii7 rather than trying to fix the json.